drivers/xen/gnttab.c

```diff
@@ -350,25 +372,32 @@ static int gnttab_init(void)
 		gnttab.gref_list[0] = gref;
 	}
 
-	for (i = 0; i < NR_GRANT_FRAMES; i++) {
+	for (i = 0; i < CONFIG_NR_GRANT_FRAMES; i++) {
 		xatp.domid = DOMID_SELF;
 		xatp.size = 0;
 		xatp.space = XENMAPSPACE_grant_table;
 		xatp.idx = i;
-		xatp.gpfn = xen_virt_to_gfn(Z_TOPLEVEL_ROM_NAME(grant_tables).phys_addr) + i;
+		xatp.gpfn = xen_virt_to_gfn(gnttab_base) + i;
 		rc = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
 		__ASSERT(!rc, "add_to_physmap failed; status = %d\n", rc);
 	}
 
 	setup.dom = DOMID_SELF;
-	setup.nr_frames = NR_GRANT_FRAMES;
+	setup.nr_frames = CONFIG_NR_GRANT_FRAMES;
 	set_xen_guest_handle(setup.frame_list, frames);
 	rc = HYPERVISOR_grant_table_op(GNTTABOP_setup_table, &setup, 1);
 	__ASSERT((!rc) && (!setup.status), "Table setup failed; status = %s\n",
 		gnttabop_error(setup.status));
 
-	DEVICE_MMIO_TOPLEVEL_MAP(grant_tables, K_MEM_CACHE_WB | K_MEM_PERM_RW);
-	gnttab.table = (grant_entry_v1_t *)DEVICE_MMIO_TOPLEVEL_GET(grant_tables);
+	/*
+	 * Xen DT region reserved for grant table (first reg in hypervisor node)
+	 * may be much bigger than CONFIG_NR_GRANT_FRAMES multiplied by page size.
+	 * Thus, we need to map only part of region, that is limited by config.
+	 * The size of this part is calculated in GNTTAB_SIZE macro and used as
+	 * parameter for device_map()
+	 */
+	device_map(&gnttab_reg, gnttab_base, GNTTAB_SIZE, K_MEM_CACHE_WB | K_MEM_PERM_RW);
```

Comment:
Just a question - does  K_MEM_CACHE_WB and caching needed here? My understanding it's some shared memory, and access rate not super high.

Reply:
The problem is not related to access rate, shared memory should be accessed with the same flags from different domains to avoid caching issues. Since it is normal memory, it is expected to use caching for it as for any other RAM addresses.

Reply:
Yeah, ARM Architecture Reference Manual has pretty good section on "Mismatching memory attributes"

Reply:
ok. noted